text/0000-command-overrides.md

```diff
@@ -0,0 +1,132 @@
+
+- Why should we do this?
+
+This addition would bring Buildpack API in parity with the OCI image spec and what is provided by Dockerfile. It would also enable use cases for overriding default arguments during runtime in the Kubernetes world. In order to achieve something like this, users have resolved to creating shims which implement this functionality. I believe that we could enable a lot more use cases out of the box without users having to implement it themselves.
+
+- What use cases does it support?
+
+  - Providing a default set of arguments
+  - Over-riding these defaults
+
+- What is the expected outcome?
+
+The launcher should be able to determine if arguments have been passed by the user or not and if they have, it should override a default list of arguments by the once specified by the user.
+
+
+# What it is
+[what-it-is]: #what-it-is
+
+The proposal essentially involves adding an additional `additional_default_args` key in the `Launch Metadata` (launch.toml) that is a list of strings. This list of strings would be a set of additional arguments that would be appended to the existing command invocation if the user does not provide any arguments themselves.
+
+```toml
+[[processes]]
+type = "<process type>"
+command = "<command>"
+args = ["<arguments>"]
+additional_default_args = ["<additional>", "<default>", "<arguments>"]
+direct = false
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Let's see how this proposal will work through an example. Let's say we have a process as defined below - 
+
+
+```toml
+[[processes]]
+type = "hi"
+command = "echo"
+args = ["Hello"]
+additional_default_args = ["world"]
+direct = <direct>
+```
+
+If the user runs - 
+
+```
+docker run --entrypoint hi <image-name>
+```
+
+The above would translate to the behavior exhibited by the following process definition as per the current spec -
+
+```toml
+[[processes]]
+type = "hi"
+command = "echo"
+args = ["Hello", "world"]
+direct = <direct>
+```
+
+On the other hand if the user provides arguments by running something like - 
+
+```
+docker run --entrypoint hi <image-name> universe
+```
+
+The above would translate to the behavior exhibited by the following process definition as per the current spec -
+
+```toml
+[[processes]]
+type = "hi"
+command = "echo"
+args = ["Hello", "universe"]
+direct = <direct>
+```
+
+This way we can preserve the current behaviour of appending user-provided arguments and remain backwards compatible, while at the same allowing a way to provide a set of additional default arguments that can be over-ridden by the user.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+N/A
+
+# Alternatives
+[alternatives]: #alternatives
+
+The other alternative to support something like this is to change `command` from a string to an array of strings specifying the command and the additional arguments it will always be run with and `args` to become a set of default arguments that can be over-ridden by the user.
```

Comment:
I think I would prefer to change `command` to match existing OCI platforms/standards.

I get the backwards compatibility concern - but is that a large concern if this is only for Buildpack API > `0.X`?  Is the concern that buildpack authors won't notice the breaking change?

This is what this alternative is proposing right? Command becoming an array and args being 100% override-able.
```toml
[[processes]]
type = "hi"
command = ["echo", "Hello"]
args = ["world"]
direct = <direct>
```

Reply:
Yup, this would have been ideal, but the only reason I didn't go with this is because it was backwards incompatible.